tianshou/env/venvs.py

```diff
@@ -0,0 +1,319 @@
+import gym
+import warnings
+import numpy as np
+from typing import List, Tuple, Union, Optional, Callable, Any
+
+from tianshou.env.worker import EnvWorker, DummyEnvWorker, SubprocEnvWorker, \
+    RayEnvWorker
+
+
+class BaseVectorEnv(gym.Env):
+    """Base class for vectorized environments wrapper. Usage:
+    ::
+
+        env_num = 8
+        envs = DummyVectorEnv([lambda: gym.make(task) for _ in range(env_num)])
+        assert len(envs) == env_num
+
+    It accepts a list of environment generators. In other words, an environment
+    generator ``efn`` of a specific task means that ``efn()`` returns the
+    environment of the given task, for example, ``gym.make(task)``.
+
+    All of the VectorEnv must inherit :class:`~tianshou.env.BaseVectorEnv`.
+    Here are some other usages:
+    ::
+
+        envs.seed(2)  # which is equal to the next line
+        envs.seed([2, 3, 4, 5, 6, 7, 8, 9])  # set specific seed for each env
+        obs = envs.reset()  # reset all environments
+        obs = envs.reset([0, 5, 7])  # reset 3 specific environments
+        obs, rew, done, info = envs.step([1] * 8)  # step synchronously
+        envs.render()  # render all environments
+        envs.close()  # close all environments
+
+    .. warning::
+
+        If you use your own environment, please make sure the ``seed`` method
+        is set up properly, e.g.,
+        ::
+
+            def seed(self, seed):
+                np.random.seed(seed)
+
+        Otherwise, the outputs of these envs may be the same with each other.
+
+    :param env_fns: a list of callable envs, ``env_fns[i]()`` will generate
+        the ith env.
+    :param worker_fn: a callable worker, ``worker_fn(env_fns[i])`` will
+        generate a worker which contains this env.
+    :param int wait_num: used in asynchronous simulation if the time cost of
+        ``env.step`` varies with time and synchronously waiting for all
+        environments to finish a step is time-wasting. In that case, we can
+        return when ``wait_num`` environments finish a step and keep on
+        simulation in these environments. If ``None``, asynchronous simulation
+        is disabled; else, ``1 <= wait_num <= env_num``.
+    """
+
+    def __init__(self,
+                 env_fns: List[Callable[[], gym.Env]],
+                 worker_fn: Callable[[Callable[[], gym.Env]], EnvWorker],
+                 wait_num: Optional[int] = None,
+                 ) -> None:
+        self._env_fns = env_fns
+        self.workers = [worker_fn(fn) for fn in env_fns]
+        self.env_num = len(env_fns)
+        self.worker_class = self.workers[0].__class__
```

Comment:
An alternative solution is to use a separate ``wait`` function and check the type of arguments to determine how to wait. But I'm afraid the type check overhead is large, because ``wait`` is a fundamental operator during async simulation (it will be called many, many times).

Reply:
Clearly. But it is not possible to invoke `self.workers[i].wait()` directly ?